android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/services/LocationService.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2021-2024 Ona Systems, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.smartregister.fhircore.engine.rulesengine.services
+
+import android.content.Context
+import android.location.Location
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationServices
+import dagger.hilt.android.qualifiers.ApplicationContext
+import java.time.Instant
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.hl7.fhir.r4.model.Resource
+import org.smartregister.fhircore.engine.util.SharedPreferenceKey
+import org.smartregister.fhircore.engine.util.SharedPreferencesHelper
+import org.smartregister.fhircore.engine.util.location.LocationUtils.getAccurateLocation
+import org.smartregister.fhircore.engine.util.location.LocationUtils.getApproximateLocation
+import org.smartregister.fhircore.engine.util.location.PermissionUtils.hasCoarseLocationPermissions
+import org.smartregister.fhircore.engine.util.location.PermissionUtils.hasFineLocationPermissions
+
+class LocationService(
+  @ApplicationContext val context: Context,
+  val sharedPreferences: SharedPreferencesHelper,
+) {
```

Comment:
```suggestion
class LocationService(
  @ApplicationContext val context: Context,
  val preferenceDataStore: PreferenceDataStore,
  val dispatcherProvider: DispatcherProvider,
) {
```

Reply:
1. Inject dispatcher provider to use.
2. Use preference data store instead of shared preferences to reduce the shared preference to data store migration in the future.

Reply:
@ellykits Thank you for the insightful feedback.

After incorporating the suggestion, I successfully integrated the dispatcher provider. This granted me access to the context, facilitating retrieval of location-coordinates data from the data store.

However, as the data comes in a flow, I needed to encase it within a coroutine context to collect it. The challenge arose when the rules were computed before we could retrieve the desired value.

To address this, I explored the option of implementing a run-blocking mechanism for a definite time. Yet, I recognized this could potentially freeze the UI.